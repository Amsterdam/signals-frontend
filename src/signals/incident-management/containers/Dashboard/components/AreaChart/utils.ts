// SPDX-License-Identifier: MPL-2.0
// Copyright (C) 2023 Gemeente Amsterdam
import { Direction } from './ComparisonRate'
import type { ComparisonRateType } from './ComparisonRate'
import type { AreaChartValue as Value, Today } from '../../charts/types'

export const formatData = (values: Value[]): Value[] => {
  return values.map((value, i) =>
    i === values.length - 1
      ? {
          ...value,
          date: `${value.date} 12:00`,
          image: '/assets/images/area-chart-tooltip.svg',
        }
      : { ...value, date: `${value.date} 12:00` }
  )
}

export const getToday = (): Today => {
  const today = new Date()

  return {
    year: today.getFullYear(),
    // Vega lite does not work with zero indexed months
    month: today.getMonth() + 1,
    date: today.getDate(),
    hours: 12,
    minutes: 0o0,
  }
}

export const getPercentage = (values: Value[]): ComparisonRateType => {
  const thisWeek = values.reduce((total, value) => total + value.amount, 0)
  const lastWeek = values.reduce(
    (total, value) => total + value.amount_week_earlier,
    0
  )

  return {
    direction: thisWeek < lastWeek ? Direction.DOWN : Direction.UP,
    percentage: Math.abs(Math.round((thisWeek / lastWeek - 1) * 100)),
  }
}
