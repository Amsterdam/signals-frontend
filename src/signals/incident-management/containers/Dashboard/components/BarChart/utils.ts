// SPDX-License-Identifier: MPL-2.0
// Copyright (C) 2023 Gemeente Amsterdam
import configuration from 'shared/services/configuration/configuration'

import type { RawData } from './types'
import { constants } from '../../charts'
import type { BarChartValue } from '../../charts'

export const getQueryList = (queryString: string) => {
  const endpoint = configuration.INCIDENT_PRIVATE_ENDPOINT
  return constants.statusList.map((status) => {
    const params = queryString
      ? `${queryString}&status=${status.query}`
      : `status=${status.query}`
    return `${endpoint}stats/total?${params}`
  })
}

export const formatData = (rawData: RawData[]): BarChartValue[] => {
  return rawData.map((value, i) => {
    return {
      status: constants.statusList[i].label,
      nrOfIncidents: value.total,
    }
  })
}

export const getTotalNrOfIncidents = (rawData: RawData[]): number => {
  return rawData.reduce(
    (aggregatedTotal, value) => aggregatedTotal + value.total,
    0
  )
}
