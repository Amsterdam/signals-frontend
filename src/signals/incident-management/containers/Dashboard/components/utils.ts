// SPDX-License-Identifier: MPL-2.0
// Copyright (C) 2023 Gemeente Amsterdam
import type { IncidentCount } from '../charts/types'

export const getMaxDomain = (values: IncidentCount[]) =>
  Math.max(...values.map((v) => v.amount)) * 1.2
