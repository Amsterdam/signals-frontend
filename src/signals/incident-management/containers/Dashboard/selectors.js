import { createSelector } from 'reselect';

/**
 * Direct selector to the dashboard state domain
 */
const selectDashboardDomain = (state) => state.get('dashboard');

/**
 * Other specific selectors
 */


/**
 * Default selector used by Dashboard
 */

const makeSelectDashboard = () => createSelector(
  selectDashboardDomain,
  (substate) => substate.toJS()
);

export default makeSelectDashboard;
export {
  selectDashboardDomain,
};
