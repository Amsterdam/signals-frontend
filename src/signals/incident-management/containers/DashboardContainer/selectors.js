import { createSelector } from 'reselect';

/**
 * Direct selector to the dashboard state domain
 */
const selectDashboardContainerDomain = (state) => state.get('incidentDashboardContainer');

/**
 * Other specific selectors
 */


/**
 * Default selector used by Dashboard
 */

const makeSelectDashboardContainer = () => createSelector(
  selectDashboardContainerDomain,
  (substate) => substate.toJS()
);

export default makeSelectDashboardContainer;
export {
  selectDashboardContainerDomain,
};
