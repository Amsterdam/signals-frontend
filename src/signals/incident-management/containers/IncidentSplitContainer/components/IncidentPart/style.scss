@import "config.scss";

.incident-part {
  border-bottom: 1px solid $ams-middengrijs;
  padding-bottom: 15px;
  margin-bottom: 37px;
}
