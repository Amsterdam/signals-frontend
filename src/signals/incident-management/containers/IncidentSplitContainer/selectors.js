import { createSelector } from 'reselect';

/**
 * Direct selector to the incidentSplitContainer state domain
 */
const selectIncidentSplitContainerDomain = state => state.get('incidentSplitContainer');

/**
 * Other specific selectors
 */


/**
 * Default selector used by IncidentSplitContainer
 */

const makeSelectIncidentSplitContainer = () => createSelector(
  selectIncidentSplitContainerDomain,
  substate => substate.toJS()
);

export default makeSelectIncidentSplitContainer;
export {
  selectIncidentSplitContainerDomain,
};
