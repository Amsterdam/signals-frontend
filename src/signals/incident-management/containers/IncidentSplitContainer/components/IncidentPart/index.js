import React, { Fragment, useMemo } from 'react';
import PropTypes from 'prop-types';
import { Heading, themeColor, themeSpacing } from '@datapunt/asc-ui';
import styled from 'styled-components';
import { useSelector } from 'react-redux';

import { makeSelectSubCategories } from 'models/categories/selectors';
import { attachmentsType } from 'shared/types';
import { priorityList, typesList } from 'signals/incident-management/definitions';

import FieldControlWrapper from 'signals/incident-management/components/FieldControlWrapper';
import CopyFileInput from 'signals/incident-management/components/CopyFileInput';
import RadioInput from 'signals/incident-management/components/RadioInput';
import SelectInput from 'signals/incident-management/components/SelectInput';
import TextAreaInput from 'signals/incident-management/components/TextAreaInput';

const StyledWrapper = styled.div`
  border-bottom: 2px solid ${themeColor('tint', 'level3')};
  padding-bottom: ${themeSpacing(4)};
  margin-bottom: ${themeSpacing(10)};
`;

const StyledH2 = styled(Heading)`
  font-weight: normal;
  margin-bottom: ${themeSpacing(3)};
`;

const IncidentPart = ({ index, attachments, splitForm }) => {
  const subcategories = useSelector(makeSelectSubCategories);
  const subcategoryOptions = useMemo(
    () =>
      subcategories?.map(category => ({
        ...category,
        value: category.extendedName,
      })),
    // disabling linter; we want to allow possible null subcategories
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [subcategories]
  );

  if (!subcategories) return null;

  return (
    <StyledWrapper>
      <StyledH2 forwardedAs="h2" data-testid="incidentPartTitle">
        Deelmelding {index}
      </StyledH2>

      {splitForm && (
        <Fragment>
          <FieldControlWrapper
            data-testid="incidentPartFieldSubcategory"
            render={SelectInput}
            name={`part${index}.subcategory`}
            display="Subcategorie"
            control={splitForm.get(`part${index}.subcategory`)}
            values={subcategoryOptions}
            sort
          />

          <FieldControlWrapper
            data-testid="incidentPartFieldText"
            render={TextAreaInput}
            name={`part${index}.text`}
            display="Omschrijving"
            control={splitForm.get(`part${index}.text`)}
            rows={5}
          />

          {attachments?.length > 0 && (
            <FieldControlWrapper
              render={CopyFileInput}
              name={`part${index}.image`}
              control={splitForm.get(`part${index}.image`)}
              values={attachments}
            />
          )}

          <FieldControlWrapper
            render={TextAreaInput}
            name={`part${index}.note`}
            display="Notitie"
            control={splitForm.get(`part${index}.note`)}
            rows={5}
          />

          <FieldControlWrapper
            render={RadioInput}
            name={`part${index}.priority`}
            display="Urgentie"
            control={splitForm.get(`part${index}.priority`)}
            values={priorityList}
          />

          <FieldControlWrapper
            render={RadioInput}
            name={`part${index}.type`}
            display="Type"
            control={splitForm.get(`part${index}.type`)}
            values={typesList}
          />
        </Fragment>
      )}
    </StyledWrapper>
  );
};

IncidentPart.defaultProps = {
  attachments: [],
  splitForm: null,
};

IncidentPart.propTypes = {
  incident: PropTypes.shape({
    category: PropTypes.shape({}),
    priority: PropTypes.shape({
      priority: PropTypes.string,
    }),
    type: PropTypes.shape({
      code: PropTypes.string.isRequired,
    }),
  }),
  index: PropTypes.string.isRequired,
  attachments: attachmentsType,
  splitForm: PropTypes.shape({
    get: PropTypes.func,
  }),
};

export default IncidentPart;
