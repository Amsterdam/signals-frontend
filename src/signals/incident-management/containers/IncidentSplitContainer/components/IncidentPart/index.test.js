import React from 'react';
import { render } from '@testing-library/react';
import { FormControl } from 'react-reactive-form';

import { withAppContext } from 'test/utils';
import incident from 'utils/__tests__/fixtures/incident.json';
import * as modelSelectors from 'models/categories/selectors';
import categoriesFixture from 'utils/__tests__/fixtures/categories_private.json';
import { filterForSub } from 'models/categories/selectors';

import IncidentPart from './index';

jest.mock('containers/App/selectors', () => ({
  __esModule: true,
  ...jest.requireActual('containers/App/selectors'),
}));

const subCategories = categoriesFixture.results
  .filter(filterForSub)
  // mapping subcategories to prevent a warning about non-unique keys rendered by the SelectInput element ðŸ™„
  .map(subCat => ({ ...subCat, key: subCat._links.self.href }));

describe('<IncidentPart />', () => {
  let props;

  beforeEach(() => {
    jest.spyOn(modelSelectors, 'makeSelectSubCategories').mockImplementation(() => subCategories);

    const splitForm = {
      get: jest.fn().mockImplementation(() => new FormControl()),
    };

    props = {
      index: '2',
      incident,
      attachments: [],
      splitForm,
    };
  });

  describe('rendering', () => {
    it('should render correctly', () => {
      const { queryByTestId, queryByText, rerender } = render(withAppContext(<IncidentPart {...props} />));

      expect(queryByTestId('incidentPartTitle')).toHaveTextContent(/^Deelmelding 2$/);
      expect(queryByText('Subcategorie')).toBeInTheDocument();
      expect(queryByText('Omschrijving')).toBeInTheDocument();
      expect(queryByText('Notitie')).toBeInTheDocument();
      expect(queryByText('Urgentie')).toBeInTheDocument();
      expect(queryByText('Type')).toBeInTheDocument();
      expect(queryByText("Foto's toevoegen")).not.toBeInTheDocument();

      props.attachments.push({ location: 'mock-image' });
      rerender(withAppContext(<IncidentPart {...props} />));

      expect(queryByText("Foto's toevoegen")).toBeInTheDocument();
    });

    it('should wait till all categories have been loaded', () => {
      const { container, rerender, unmount } = render(withAppContext(<IncidentPart {...props} />));

      expect(container.querySelector('select')).toBeInTheDocument();

      jest.spyOn(modelSelectors, 'makeSelectSubCategories').mockImplementation(() => null);

      unmount();

      rerender(withAppContext(<IncidentPart {...props} />));

      expect(container.querySelector('select')).not.toBeInTheDocument();
    });

    it('should render no form components when form is not available', () => {
      props.splitForm = undefined;
      const { queryByTestId, queryByText } = render(withAppContext(<IncidentPart {...props} />));

      expect(queryByTestId('incidentPartTitle')).toBeInTheDocument();
      expect(queryByText('Subcategorie')).not.toBeInTheDocument();
      expect(queryByText('Omschrijving')).not.toBeInTheDocument();
      expect(queryByText('Notitie')).not.toBeInTheDocument();
      expect(queryByText('Urgentie')).not.toBeInTheDocument();
      expect(queryByText('Type')).not.toBeInTheDocument();
      expect(queryByText("Foto's toevoegen")).not.toBeInTheDocument();
    });
  });
});
