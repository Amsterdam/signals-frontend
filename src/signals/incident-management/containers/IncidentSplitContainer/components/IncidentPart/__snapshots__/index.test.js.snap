// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IncidentPart /> rendering should render correctly with image 1`] = `
<section
  className="incident-part"
>
  <h2>
    Deelmelding
    2
  </h2>
  <FieldControlWrapper
    control={
      Object {
        "item": "part2.subcategory",
      }
    }
    display="Subcategorie"
    emptyOptionText=""
    name="part2.subcategory"
    render={[Function]}
    sort={true}
    values={
      Array [
        Object {
          "key": "key",
          "slug": "slug",
          "value": "value",
        },
      ]
    }
  />
  <FieldControlWrapper
    control={
      Object {
        "item": "part2.text",
      }
    }
    display="Omschrijving"
    emptyOptionText=""
    name="part2.text"
    render={[Function]}
    rows={5}
    values={Array []}
  />
  <FieldControlWrapper
    control={
      Object {
        "item": "part2.image",
      }
    }
    emptyOptionText=""
    name="part2.image"
    render={[Function]}
    values={
      Array [
        Object {
          "location": "mock-image",
        },
      ]
    }
  />
  <FieldControlWrapper
    control={
      Object {
        "item": "part2.note",
      }
    }
    display="Notitie"
    emptyOptionText=""
    name="part2.note"
    render={[Function]}
    rows={5}
    values={Array []}
  />
  <FieldControlWrapper
    control={
      Object {
        "item": "part2.priority",
      }
    }
    display="Urgentie"
    emptyOptionText=""
    name="part2.priority"
    render={[Function]}
    values={
      Array [
        Object {
          "key": "normal",
          "value": "Normaal",
        },
        Object {
          "key": "high",
          "value": "Hoog",
        },
      ]
    }
  />
</section>
`;

exports[`<IncidentPart /> rendering should render correctly without image 1`] = `
<section
  className="incident-part"
>
  <h2>
    Deelmelding
    2
  </h2>
  <FieldControlWrapper
    control={
      Object {
        "item": "part2.subcategory",
      }
    }
    display="Subcategorie"
    emptyOptionText=""
    name="part2.subcategory"
    render={[Function]}
    sort={true}
    values={
      Array [
        Object {
          "key": "key",
          "slug": "slug",
          "value": "value",
        },
      ]
    }
  />
  <FieldControlWrapper
    control={
      Object {
        "item": "part2.text",
      }
    }
    display="Omschrijving"
    emptyOptionText=""
    name="part2.text"
    render={[Function]}
    rows={5}
    values={Array []}
  />
  0
  <FieldControlWrapper
    control={
      Object {
        "item": "part2.note",
      }
    }
    display="Notitie"
    emptyOptionText=""
    name="part2.note"
    render={[Function]}
    rows={5}
    values={Array []}
  />
  <FieldControlWrapper
    control={
      Object {
        "item": "part2.priority",
      }
    }
    display="Urgentie"
    emptyOptionText=""
    name="part2.priority"
    render={[Function]}
    values={
      Array [
        Object {
          "key": "normal",
          "value": "Normaal",
        },
        Object {
          "key": "high",
          "value": "Hoog",
        },
      ]
    }
  />
</section>
`;
